File: windows/Asr/FbankFrontend.cs
using System;
using System.Collections.Generic;

namespace VoiceTyper.Asr;

/// <summary>
/// Kaldi 兼容的 fbank 特征前端。C# 直译自 <c>macos/Sources/VoiceTyper/ASR/FbankFrontend.swift</c>，
/// 参数对齐 <c>kaldi_native_fbank</c>（knf 1.22.3）在 <c>client-server/server/voice_typer_server/recognizer.py</c>
/// 里的生效配置：
///
/// <code>
/// frame_opts: samp_freq=16000  frame_length_ms=25  frame_shift_ms=10
///             dither=0  preemph_coeff=0.97  remove_dc_offset=true
///             window_type=hamming  round_to_power_of_two=true  snip_edges=true
/// mel_opts:   num_bins=80  low_freq=20  high_freq=0(→nyquist)  is_librosa=0
/// </code>
///
/// 用 <c>macos/Tests/VoiceTyperTests/Fixtures/</c> 下的金标准夹具逐帧比对验证（阈值 1e-3），
/// 与 macOS 侧共用同一份 Python 参考输出。
/// </summary>
internal sealed class FbankOptions
{
    public double SampleRate { get; init; } = 16000;
    public double FrameLengthMs { get; init; } = 25;
    public double FrameShiftMs { get; init; } = 10;
    public int NumMelBins { get; init; } = 80;
    public double LowFreq { get; init; } = 20;
    /// <summary>0 表示用 Nyquist（sampleRate/2）。</summary>
    public double HighFreq { get; init; } = 0;
    public double PreemphCoeff { get; init; } = 0.97;
    public bool RemoveDCOffset { get; init; } = true;
}

internal sealed class FbankFrontend
{
    /// <summary>
    /// 不足一帧（25ms @16kHz = 400 samples）的输入无法产出任何 fbank 帧。
    /// 对应服务端 recognizer.py 的 <c>_MIN_SAMPLES = 400</c> 硬下限。
    /// </summary>
    public const int MinimumSamples = 400;

    /// <summary>
    /// Kaldi 的能量下限用最小正规格化 float，不是 <see cref="float.Epsilon"/>（次正规数，
    /// 差 7 个数量级）。写错不会崩，只会让静音帧的 log 能量偏低几十——这是最难在测试里
    /// 发现的一类 bug，金标准测试的静音段能抓到。
    /// </summary>
    private const float FloatMin = 1.17549435E-38f;

    private readonly FbankOptions _opts;
    public int FrameLength { get; }
    public int FrameShift { get; }
    private readonly int _fftSize;
    private readonly float[] _window;
    private readonly float[][] _melFilters;
    private readonly (int lo, int hi)[] _melBinRanges;
    private readonly Rfft _fft;

    public FbankFrontend(FbankOptions? opts = null)
    {
        _opts = opts ?? new FbankOptions();

        FrameLength = (int)Math.Round(_opts.FrameLengthMs * _opts.SampleRate / 1000);
        FrameShift = (int)Math.Round(_opts.FrameShiftMs * _opts.SampleRate / 1000);

        int n = 1;
        while (n < FrameLength) n <<= 1; // round_to_power_of_two
        _fftSize = n;
        _fft = new Rfft(n);

        // Hamming 窗，作用在原始 FrameLength 个样本上（不是补零后的 fftSize）。
        _window = new float[FrameLength];
        double denom = FrameLength - 1;
        for (int i = 0; i < FrameLength; i++)
        {
            _window[i] = (float)(0.54 - 0.46 * Math.Cos(2.0 * Math.PI * i / denom));
        }

        double nyquist = _opts.SampleRate / 2;
        double highFreq = _opts.HighFreq > 0 ? _opts.HighFreq : nyquist;
        double MelScale(double freq) => 1127.0 * Math.Log(1.0 + freq / 700.0);
        double melLow = MelScale(_opts.LowFreq);
        double melHigh = MelScale(highFreq);
        int numBins = _opts.NumMelBins;
        double melFreqDelta = (melHigh - melLow) / (numBins + 1);

        int numFFTBins = _fftSize / 2;
        double fftBinWidth = _opts.SampleRate / _fftSize;

        var filters = new List<float[]>(numBins);
        var ranges = new (int, int)[numBins];
        for (int bin = 0; bin < numBins; bin++)
        {
            double leftMel = melLow + bin * melFreqDelta;
            double centerMel = melLow + (bin + 1) * melFreqDelta;
            double rightMel = melLow + (bin + 2) * melFreqDelta;

            var thisBin = new float[numFFTBins];
            int firstIndex = -1;
            int lastIndex = -1;
            for (int fftBin = 0; fftBin < numFFTBins; fftBin++)
            {
                double freq = fftBin * fftBinWidth;
                double mel = MelScale(freq);
                if (!(mel > leftMel && mel < rightMel)) continue;
                double weight = mel <= centerMel
                    ? (mel - leftMel) / (centerMel - leftMel)
                    : (rightMel - mel) / (rightMel - centerMel);
                thisBin[fftBin] = (float)weight;
                if (firstIndex == -1) firstIndex = fftBin;
                lastIndex = fftBin;
            }
            if (firstIndex == -1) { firstIndex = 0; lastIndex = -1; }

            int span = Math.Max(0, lastIndex + 1 - firstIndex);
            var dense = new float[span];
            Array.Copy(thisBin, firstIndex, dense, 0, span);
            filters.Add(dense);
            ranges[bin] = (firstIndex, lastIndex);
        }
        _melFilters = filters.ToArray();
        _melBinRanges = ranges;
    }

    /// <param name="waveform">原始 float32 采样，范围 [-1, 1]（尚未按 knf 约定乘以 32768）。</param>
    /// <returns><c>[numFrames][numMelBins]</c> 的对数 mel 能量；样本不足一帧时返回空数组。</returns>
    public float[][] Compute(ReadOnlySpan<float> waveform)
    {
        int n = waveform.Length;
        if (n < FrameLength) return Array.Empty<float[]>();

        var scaled = new float[n];
        for (int i = 0; i < n; i++) scaled[i] = waveform[i] * 32768.0f;

        int numFrames = (n - FrameLength) / FrameShift + 1;
        var result = new float[numFrames][];

        var real = new float[_fftSize];
        var imag = new float[_fftSize];

        for (int f = 0; f < numFrames; f++)
        {
            int start = f * FrameShift;
            for (int i = 0; i < FrameLength; i++) real[i] = scaled[start + i];
            for (int i = FrameLength; i < _fftSize; i++) real[i] = 0f;
            Array.Clear(imag, 0, _fftSize);

            if (_opts.RemoveDCOffset)
            {
                float mean = 0;
                for (int i = 0; i < FrameLength; i++) mean += real[i];
                mean /= FrameLength;
                for (int i = 0; i < FrameLength; i++) real[i] -= mean;
            }

            if (_opts.PreemphCoeff != 0)
            {
                // Kaldi 的预加重按倒序原地修改：x[i] -= coeff*x[i-1]（i 从高到低），
                // 最后单独处理 x[0] -= coeff*x[0]。顺序不可颠倒，否则会用到"未来"的已修改值。
                float coeff = (float)_opts.PreemphCoeff;
                for (int i = FrameLength - 1; i > 0; i--)
                {
                    real[i] -= coeff * real[i - 1];
                }
                real[0] -= coeff * real[0];
            }

            for (int i = 0; i < FrameLength; i++) real[i] *= _window[i];

            _fft.Forward(real, imag);

            var melEnergies = new float[_opts.NumMelBins];
            int numFFTBins = _fftSize / 2;
            for (int bin = 0; bin < _opts.NumMelBins; bin++)
            {
                var (lo, hi) = _melBinRanges[bin];
                if (hi < lo) continue;
                var weights = _melFilters[bin];
                float sum = 0;
                for (int k = lo; k <= hi; k++)
                {
                    float power = real[k] * real[k] + imag[k] * imag[k];
                    sum += power * weights[k - lo];
                }
                melEnergies[bin] = MathF.Log(MathF.Max(sum, FloatMin));
            }
            result[f] = melEnergies;
        }
        return result;
    }
}

File: macos/Sources/VoiceTyper/ASR/FbankFrontend.swift
import Accelerate
import Foundation

/// Kaldi 兼容的 fbank 特征前端，参数对齐 `kaldi_native_fbank`（knf 1.22.3）在
/// `client-server/server/voice_typer_server/recognizer.py` 里的生效配置：
///
/// ```
/// frame_opts: samp_freq=16000  frame_length_ms=25  frame_shift_ms=10
///             dither=0  preemph_coeff=0.97  remove_dc_offset=true
///             window_type=hamming  round_to_power_of_two=true  snip_edges=true
/// mel_opts:   num_bins=80  low_freq=20  high_freq=0(→nyquist)  is_librosa=0
/// ```
///
/// 已用 `dump_reference_fixtures.py` 生成的金标准逐帧比对：max abs diff ≈ 6e-5
/// （详见 macos/DESIGN.md §4.2、FbankParityTests）。
struct FbankOptions {
    var sampleRate: Double = 16000
    var frameLengthMs: Double = 25
    var frameShiftMs: Double = 10
    var numMelBins: Int = 80
    var lowFreq: Double = 20
    /// 0 表示用 Nyquist（sampleRate/2）。
    var highFreq: Double = 0
    var preemphCoeff: Double = 0.97
    var removeDCOffset: Bool = true
}

/// 不足一帧（25ms @16kHz = 400 samples）的输入无法产出任何 fbank 帧。
/// 对应服务端 recognizer.py 的 `_MIN_SAMPLES = 400` 硬下限。
let fbankMinimumSamples = 400

final class FbankFrontend {
    private let opts: FbankOptions
    let frameLength: Int
    let frameShift: Int
    private let fftSize: Int
    private let window: [Float]
    /// 每个 mel bin 只在 [lo, hi] 范围内的 FFT bin 上有非零权重；稠密存储该子区间即可。
    private let melFilters: [[Float]]
    private let melBinRanges: [(lo: Int, hi: Int)]
    private let log2n: vDSP_Length
    private let fftSetup: FFTSetup

    init(opts: FbankOptions = FbankOptions()) {
        self.opts = opts
        self.frameLength = Int((opts.frameLengthMs * opts.sampleRate / 1000).rounded())
        self.frameShift = Int((opts.frameShiftMs * opts.sampleRate / 1000).rounded())

        var n = 1
        while n < frameLength { n <<= 1 } // round_to_power_of_two
        self.fftSize = n
        self.log2n = vDSP_Length(log2(Double(n)))
        self.fftSetup = vDSP_create_fftsetup(log2n, FFTRadix(kFFTRadix2))!

        // Hamming 窗，作用在原始 frameLength 个样本上（不是补零后的 fftSize）。
        var w = [Float](repeating: 0, count: frameLength)
        let denom = Double(frameLength - 1)
        for i in 0..<frameLength {
            w[i] = Float(0.54 - 0.46 * cos(2.0 * Double.pi * Double(i) / denom))
        }
        self.window = w

        let nyquist = opts.sampleRate / 2
        let highFreq = opts.highFreq > 0 ? opts.highFreq : nyquist
        func melScale(_ freq: Double) -> Double { 1127.0 * log(1.0 + freq / 700.0) }
        let melLow = melScale(opts.lowFreq)
        let melHigh = melScale(highFreq)
        let numBins = opts.numMelBins
        let melFreqDelta = (melHigh - melLow) / Double(numBins + 1)

        let numFFTBins = fftSize / 2
        let fftBinWidth = opts.sampleRate / Double(fftSize)

        var filters: [[Float]] = []
        var ranges: [(Int, Int)] = []
        for bin in 0..<numBins {
            let leftMel = melLow + Double(bin) * melFreqDelta
            let centerMel = melLow + Double(bin + 1) * melFreqDelta
            let rightMel = melLow + Double(bin + 2) * melFreqDelta

            var thisBin = [Float](repeating: 0, count: numFFTBins + 1)
            var firstIndex = -1
            var lastIndex = -1
            for fftBin in 0..<numFFTBins {
                let freq = Double(fftBin) * fftBinWidth
                let mel = melScale(freq)
                guard mel > leftMel, mel < rightMel else { continue }
                let weight = mel <= centerMel
                    ? (mel - leftMel) / (centerMel - leftMel)
                    : (rightMel - mel) / (rightMel - centerMel)
                thisBin[fftBin] = Float(weight)
                if firstIndex == -1 { firstIndex = fftBin }
                lastIndex = fftBin
            }
            if firstIndex == -1 { firstIndex = 0; lastIndex = -1 }
            filters.append(Array(thisBin[firstIndex..<max(firstIndex, lastIndex + 1)]))
            ranges.append((firstIndex, lastIndex))
        }
        self.melFilters = filters
        self.melBinRanges = ranges
    }

    deinit {
        vDSP_destroy_fftsetup(fftSetup)
    }

    /// - Parameter waveform: 原始 float32 采样，范围 [-1, 1]（尚未按 knf 约定乘以 32768）。
    /// - Returns: `[numFrames][numMelBins]` 的对数 mel 能量；样本不足一帧时返回空数组。
    func compute(_ waveform: [Float]) -> [[Float]] {
        let n = waveform.count
        guard n >= frameLength else { return [] }

        let scaled = waveform.map { $0 * 32768.0 }
        let numFrames = (n - frameLength) / frameShift + 1

        var result: [[Float]] = []
        result.reserveCapacity(numFrames)

        var frameBuf = [Float](repeating: 0, count: fftSize)
        for f in 0..<numFrames {
            let start = f * frameShift
            for i in 0..<frameLength { frameBuf[i] = scaled[start + i] }
            for i in frameLength..<fftSize { frameBuf[i] = 0 }

            if opts.removeDCOffset {
                var mean: Float = 0
                vDSP_meanv(frameBuf, 1, &mean, vDSP_Length(frameLength))
                var negMean = -mean
                vDSP_vsadd(frameBuf, 1, &negMean, &frameBuf, 1, vDSP_Length(frameLength))
            }

            if opts.preemphCoeff != 0 {
                // Kaldi 的预加重按倒序原地修改：x[i] -= coeff*x[i-1]（i 从高到低），
                // 最后单独处理 x[0] -= coeff*x[0]。顺序不可颠倒，否则会用到"未来"的已修改值。
                let coeff = Float(opts.preemphCoeff)
                var i = frameLength - 1
                while i > 0 {
                    frameBuf[i] -= coeff * frameBuf[i - 1]
                    i -= 1
                }
                frameBuf[0] -= coeff * frameBuf[0]
            }

            for i in 0..<frameLength { frameBuf[i] *= window[i] }

            let power = powerSpectrum(frameBuf)
            // 用 vDSP_dotpr 做每个 mel bin 的加权求和（替代标量循环），并把 80 个 bin
            // 的 log 一次性交给 vvlogf 批量处理（替代逐个调用 logf），减少每帧的标量
            // 循环与函数调用开销——120s 音频对应约 12000 帧，累计可观（R3-08）。
            var melEnergies = [Float](repeating: 0, count: opts.numMelBins)
            power.withUnsafeBufferPointer { powerPtr in
                for bin in 0..<opts.numMelBins {
                    let (lo, hi) = melBinRanges[bin]
                    guard hi >= lo else { continue }
                    let weights = melFilters[bin]
                    var sum: Float = 0
                    weights.withUnsafeBufferPointer { weightPtr in
                        vDSP_dotpr(powerPtr.baseAddress! + lo, 1, weightPtr.baseAddress!, 1, &sum, vDSP_Length(hi - lo + 1))
                    }
                    melEnergies[bin] = max(sum, Float.leastNormalMagnitude)
                }
            }
            var logged = [Float](repeating: 0, count: opts.numMelBins)
            melEnergies.withUnsafeBufferPointer { inPtr in
                logged.withUnsafeMutableBufferPointer { outPtr in
                    var count = Int32(opts.numMelBins)
                    vvlogf(outPtr.baseAddress!, inPtr.baseAddress!, &count)
                }
            }
            result.append(logged)
        }
        return result
    }

    /// 实数输入的功率谱，vDSP 实 FFT，返回 bin 0…fftSize/2（含 Nyquist）。
    private func powerSpectrum(_ frame: [Float]) -> [Float] {
        let half = fftSize / 2
        var realp = [Float](repeating: 0, count: half)
        var imagp = [Float](repeating: 0, count: half)
        var power = [Float](repeating: 0, count: half + 1)

        realp.withUnsafeMutableBufferPointer { realPtr in
            imagp.withUnsafeMutableBufferPointer { imagPtr in
                var split = DSPSplitComplex(realp: realPtr.baseAddress!, imagp: imagPtr.baseAddress!)
                frame.withUnsafeBufferPointer { framePtr in
                    framePtr.baseAddress!.withMemoryRebound(to: DSPComplex.self, capacity: half) { complexPtr in
                        vDSP_ctoz(complexPtr, 2, &split, 1, vDSP_Length(half))
                    }
                }
                vDSP_fft_zrip(fftSetup, &split, 1, log2n, FFTDirection(kFFTDirection_Forward))

                // vDSP_fft_zrip 的输出是"标准"正向 DFT 结果的 2 倍；实数 FFT 打包约定下
                // realp[0] 存 DC 分量（bin 0），imagp[0] 存 Nyquist 分量（bin fftSize/2）。
                let scale: Float = 0.5
                let dc = split.realp[0] * scale
                let nyq = split.imagp[0] * scale
                power[0] = dc * dc
                power[half] = nyq * nyq
                for k in 1..<half {
                    let re = split.realp[k] * scale
                    let im = split.imagp[k] * scale
                    power[k] = re * re + im * im
                }
            }
        }
        return power
    }
}
